python/bindings/openravepy_int.cpp

```diff
@@ -124,40 +124,67 @@ object toPyObject(const rapidjson::Value& value)
     }
 }
 
-#define FILL_RAPIDJSON_FROMARRAY_1D(pyarrayvalues, T, rapidjsonsetfn) {  \
-        const T *vdata = reinterpret_cast<T*>(PyArray_DATA(pyarrayvalues)); \
-        for (int i = 0; i < dims[0]; i++) { \
-            rapidjson::Value elementValue; \
-            elementValue.rapidjsonsetfn(vdata[i]); \
-            value.PushBack(elementValue, allocator); \
-        } \
-} \
-
-#define FILL_RAPIDJSON_FROMARRAY_2D(pyarrayvalues, T, rapidjsonsetfn) {                            \
-        const T *vdata = reinterpret_cast<T*>(PyArray_DATA(pyarrayvalues)); \
-        for (int i = 0; i < dims[0]; i++) { \
-            rapidjson::Value colvalues(rapidjson::kArrayType); \
-            for (int j = 0; j < dims[1]; j++) { \
-                rapidjson::Value elementValue; \
-                elementValue.rapidjsonsetfn(vdata[i*dims[0]+j]); \
-                colvalues.PushBack(elementValue, allocator); \
-            } \
-            value.PushBack(colvalues, allocator); \
-        } \
-} \
-
-#define FILL_RAPIDJSON_FROMARRAY(pyarrayvalues, T, rapidjsonsetfn, ndims) { \
-        if( ndims == 1 ) { \
-            FILL_RAPIDJSON_FROMARRAY_1D(pyarrayvalues, T, rapidjsonsetfn); \
-        } \
-        else if( ndims == 2 ) { \
-            FILL_RAPIDJSON_FROMARRAY_2D(pyarrayvalues, T, rapidjsonsetfn); \
-        } \
-        else { \
-            throw OPENRAVE_EXCEPTION_FORMAT(_("do not support array object with %d dims"), ndims, ORE_InvalidArguments); \
-        } \
+using RapidJsonGenericValueUTF8 = rapidjson::GenericValue<rapidjson::UTF8<> >;
+template <typename T>
+using RapidJsonSetFn = RapidJsonGenericValueUTF8&(RapidJsonGenericValueUTF8::*)(T);
+
+template <typename T>
+void FillRapidJsonFromArray1D(PyArrayObject* pyarr,
+                              rapidjson::Value &value,
+                              rapidjson::Document::AllocatorType& allocator,
+                              RapidJsonSetFn<T> f,
+                              npy_intp const* dims)
+{ 
+    const T *pv = reinterpret_cast<T*>(PyArray_DATA(pyarr));
+    const size_t numel = dims[0];
+    value.Reserve(numel, allocator);
+    for (size_t i = 0; i < numel; ++i) { 
+        rapidjson::Value elementValue; 
+        (elementValue.*f)(pv[i]); 
+        value.PushBack(elementValue, allocator); 
+    }
+}
+
+template <typename T>
+void FillRapidJsonFromArray2D(PyArrayObject* pyarr,
+                              rapidjson::Value &value,
+                              rapidjson::Document::AllocatorType& allocator,
+                              RapidJsonSetFn<T> f,
+                              npy_intp const* dims) 
+{ 
+    const T *pv = reinterpret_cast<T*>(PyArray_DATA(pyarr));
+    const size_t numel = dims[0] * dims[1];
+    value.Reserve(numel, allocator);
```

Comment:
Is this json supposed to be `[[1,2, 3], [4, 5, 6], [7, 8, 9]]`? If so, why `Reserve` total number of elements in the first level of array?

Reply:
oops, you are right. just patched